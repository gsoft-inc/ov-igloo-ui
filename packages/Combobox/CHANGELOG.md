# @igloo-ui/combobox

## 2.1.5

### Patch Changes

- 7d36815: Updated hopper token versions and igloo icon versions.
- Updated dependencies [7d36815]
- Updated dependencies [7d36815]
  - @igloo-ui/dropdown@1.6.5
  - @igloo-ui/icon-button@1.1.6
  - @igloo-ui/list@0.5.4
  - @igloo-ui/tooltip@3.4.3
  - @igloo-ui/input@2.1.7

## 2.1.4

### Patch Changes

- 39a1d18: Fix for minor issues
- Updated dependencies [39a1d18]
  - @igloo-ui/input@2.1.6

## 2.1.3

### Patch Changes

- 5a9444f: Fix for focus state made to be on par with Orbiter/Hopper

## 2.1.2

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency
- Updated dependencies [a2df55d]
  - @igloo-ui/dropdown@1.6.4
  - @igloo-ui/icon-button@1.1.5
  - @igloo-ui/input@2.1.5
  - @igloo-ui/list@0.5.2
  - @igloo-ui/tooltip@3.4.2

## 2.1.1

### Patch Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.
- Updated dependencies [9739155]
  - @igloo-ui/tooltip@3.4.0
  - @igloo-ui/dropdown@1.6.3
  - @igloo-ui/icon-button@1.1.4
  - @igloo-ui/input@2.1.4
  - @igloo-ui/list@0.5.1

## 2.1.0

### Minor Changes

- d9f9f12: Gave the Combobox component the new Brand

### Patch Changes

- @igloo-ui/icon-button@1.1.3

## 2.0.3

### Patch Changes

- Updated dependencies [5402bcc]
  - @igloo-ui/icon-button@1.1.0
  - @igloo-ui/list@0.5.0

## 2.0.2

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.
- Updated dependencies [bf4ddef]
  - @igloo-ui/dropdown@1.6.1
  - @igloo-ui/icon-button@1.0.4
  - @igloo-ui/input@2.1.2
  - @igloo-ui/list@0.4.3
  - @igloo-ui/tooltip@3.1.6

## 2.0.1

### Patch Changes

- 1a85f2a: Fixed the Combobox so that the loading shows up if it's true, even if there are results.

## 2.0.0

### Major Changes

- 1a5bb05: Added onAfterClose, onInput, onScrollEnd and scrollEndThreshold props to the Combobox.

  ## BREAKING CHANGE

  The search no longer does the filtering itself. The user must use the onInput callback to filter the results as well as add any debouncing needed. Since it is the user filtering the results, they must also reset the results in the onAfterClose callback.

### Patch Changes

- Updated dependencies [1a5bb05]
- Updated dependencies [1a5bb05]
  - @igloo-ui/dropdown@1.6.0
  - @igloo-ui/list@0.4.2

## 1.2.1

### Patch Changes

- c3efaf5: Fixed tests and snapshots since react-aria generated ids changed and viewBox was added to icons.
- Updated dependencies [c3efaf5]
  - @igloo-ui/icon-button@1.0.3
  - @igloo-ui/input@2.1.1
  - @igloo-ui/list@0.4.1

## 1.2.0

### Minor Changes

- 9d9d72a: Allow Combobox, Select, TagPicker and Dropdown components to scroll if it's too long and runs out of room in the viewport. New properties added to Dropdown are: isReferenceWidth, isScrollable and footer. The display of the footer was moved to the dropdown instead of leaving it in combobox because the dropdown handles the scroll styles and the footer is sticky.

### Patch Changes

- Updated dependencies [9d9d72a]
- Updated dependencies [9d9d72a]
  - @igloo-ui/tooltip@3.1.4
  - @igloo-ui/dropdown@1.4.0

## 1.1.0

### Minor Changes

- 9b707d1: Added a loading feature to List, Combobox, Select and Tagpicker components.

### Patch Changes

- Updated dependencies [9b707d1]
  - @igloo-ui/list@0.4.0

## 1.0.0

### Major Changes

- ffeaeb7: Added a search icon with the showSearchIcon prop to the combobox. Also fixed keyboard navigation when searchbox is visible.

  ## BREAKING CHANGE

  The search icon is displayed by default beside the search field. To hide it, the user must add showSearch="false".

### Patch Changes

- @igloo-ui/list@0.3.1

## 0.3.0

### Minor Changes

- 0482ec6: Added action prop and updated description type to ReactNode for ListItem. Added footer and a way for the user to open and close the combobox through props.

### Patch Changes

- Updated dependencies [0482ec6]
  - @igloo-ui/list@0.3.0

## 0.2.1

### Patch Changes

- 08a062e: update packages
- Updated dependencies [6e2e96a]
- Updated dependencies [08a062e]
- Updated dependencies [6e2e96a]
- Updated dependencies [08a062e]
  - @igloo-ui/dropdown@1.3.0
  - @igloo-ui/list@0.2.1

## 0.2.0

### Minor Changes

- 4ff1942: Added new optional props to List called 'disableTabbing' and 'onOptionBlur'. Also fixed some styling for the List component.

### Patch Changes

- Updated dependencies [4ff1942]
- Updated dependencies [0b8a62f]
  - @igloo-ui/list@0.2.0
  - @igloo-ui/dropdown@1.2.0

## 0.1.7

### Patch Changes

- 84e2d78: Adding a hover effect
- Updated dependencies [84e2d78]
  - @igloo-ui/input@2.0.5

## 0.1.6

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.
- Updated dependencies [a41e1d4]
  - @igloo-ui/dropdown@1.1.4
  - @igloo-ui/icon-button@1.0.1
  - @igloo-ui/input@2.0.4
  - @igloo-ui/list@0.1.3
  - @igloo-ui/tooltip@3.1.3

## 0.1.5

### Patch Changes

- bf478b5: Added missing tests. Fixed missing code in stories when clicking 'Show code'. Updated Dropdown mock. Added new optional prop highlightToday to help with test snapshots.
- Updated dependencies [bf478b5]
  - @igloo-ui/dropdown@1.1.3
  - @igloo-ui/tooltip@3.1.2

## 0.1.4

### Patch Changes

- 55d12e3: fix error with empty package @shared/components
- Updated dependencies [55d12e3]
- Updated dependencies [49d3c6b]
  - @igloo-ui/list@0.1.2
  - @igloo-ui/tooltip@3.1.1

## 0.1.3

### Patch Changes

- 3f50827: Made dropdown portal to the end of the document so that it works in Modals. Removed unused line in Modal.
- Updated dependencies [3f50827]
  - @igloo-ui/dropdown@1.1.2

## 0.1.2

### Patch Changes

- 4230499: Fix: updated components to use the IconButton with appearance ghost, variant secondary so that they are grey like they used to be.
- Updated dependencies [4230499]
  - @igloo-ui/tooltip@3.0.3

## 0.1.1

### Patch Changes

- d5c91f5: link shared components with aliases
- Updated dependencies [e47ae88]
- Updated dependencies [d5c91f5]
  - @igloo-ui/icon-button@1.0.0
  - @igloo-ui/list@0.1.1

## 0.1.0

### Minor Changes

- 7dcba93: Initial release of List and Combobox components. Select was updated to use the Dropdown component that was created previously as well as the new List component.

### Patch Changes

- Updated dependencies [7dcba93]
  - @igloo-ui/list@0.1.0
  - @igloo-ui/dropdown@1.1.1
