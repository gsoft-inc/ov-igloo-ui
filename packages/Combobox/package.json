{
    "name": "@igloo-ui/combobox",
    "version": "2.2.2",
    "main": "dist/Combobox.js",
    "module": "dist/Combobox.js",
    "types": "dist/Combobox.d.ts",
    "styles": "dist/combobox.css",
    "publishConfig": {
        "access": "public"
    },
    "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
    "author": "Officevibe Inc.",
    "contributors": [
        "Vicky Comeau vicky.comeau@gsoft.com"
    ],
    "license": "Apache-2.0",
    "peerDependencies": {
        "react": ">= 16.8.6"
    },
    "files": [
        "dist"
    ],
    "scripts": {
        "build": "rollup -c rollup.config.js"
    },
    "dependencies": {
        "@igloo-ui/dropdown": "^1.8.0",
        "@igloo-ui/list": "^0.6.1",
        "@igloo-ui/icons": "^1.12.2",
        "@igloo-ui/icon-button": "^1.2.0",
        "@igloo-ui/input": "^2.2.1",
        "@igloo-ui/tokens": "^2.0.0",
        "@hopper-ui/icons-react16": "1.2.0",
        "@hopper-ui/tokens": "^3.1.3",
        "@igloo-ui/tooltip": "^3.5.0",
        "classnames": "^2.3.2"
    },
    "devDependencies": {
        "@igloo-ui/button": "^0.8.5",
        "@igloo-ui/provider": "^0.1.1",
        "@igloo-ui/tag": "^1.5.5",
        "@shared/components": "0.5.0"
    },
    "browserslist": [
        "> 1%",
        "last 2 versions",
        "IE >= 11"
    ]
}
