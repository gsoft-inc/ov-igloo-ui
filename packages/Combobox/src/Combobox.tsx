import * as React from 'react';
import cx from 'classnames';

import Dropdown from '@igloo-ui/dropdown';
import List, { OptionType, Option } from '@igloo-ui/list';

import ComboboxInput from './ComboboxInput';

import './combobox.scss';

export enum Keys {
  Enter = 'Enter',
  Space = ' ',
  ArrowDown = 'ArrowDown',
  ArrowUp = 'ArrowUp',
  Escape = 'Escape',
  Tab = 'Tab',
}

export type FocusDirection = 'first' | 'last' | 'up' | 'down';

export type ComboboxOption = Omit<Option, 'type'>;

export interface ComboboxProps
  extends Omit<React.ComponentProps<'div'>, 'onChange' | 'onInput'> {
  /** Set this to true and the dropdown will take the width of its content,
   * not the width of the select */
  autoWidth?: boolean;
  /** Default value displayed in the Select */
  children?: React.ReactNode;
  /** Add a specific class to the Select */
  className?: string;
  /** If true, it clears any selected option */
  clear?: boolean;
  /** The tooltip text for the clear button */
  clearTooltipText?: string;
  /** Whether or not the combobox should close after an item is selected */
  closeOnSelect?: boolean;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Disable the Select so the user cannot click on it */
  disabled?: boolean;
  /** The Select is in an error state */
  error?: boolean;
  /** The footer content of the combobox */
  footer?: React.ReactNode;
  /** True for a compact appearance */
  isCompact?: boolean;
  /** True if the option list is displayed */
  isOpen?: boolean;
  /** Whether or not the list is loading */
  loading?: boolean;
  /** The Combobox gains checkboxes beside each option
   * to be able to select multiple options */
  multiple?: boolean;
  /** Specify the text to display when there are no results found */
  noResultsText?: string;
  /** Callback when the dropdown is closed and animations are done */
  onAfterClose?: () => void;
  /** Callback when selected content changes */
  onChange?: (option: OptionType | undefined) => void;
  /** Callback called when selected is cleared */
  onClear?: () => void;
  /** Callback when the dropdown is closed */
  onClose?: () => void;
  /** Callback when the user types in the search box */
  onInput?: (value: string) => void;
  /** Callback when the dropdown is opened */
  onOpen?: () => void;
  /** Callback when the user scrolls to the end of the dropdown */
  onScrollEnd?: () => void;
  /** List of available options. */
  options?: ComboboxOption[];
  /** The threshold in pixels from the bottom of the dropdown
   * to trigger the onScrollEnd callback */
  scrollEndThreshold?: number;
  /** Whether or not to display a search box when open */
  search?: boolean;
  /** The initial selected option or a list of selected options */
  selectedOption?: OptionType | OptionType[];
  /** Whether or not to display the search icon */
  showSearchIcon?: boolean;
}

const Combobox: React.FunctionComponent<ComboboxProps> = ({
  autoWidth = false,
  children,
  className,
  clear,
  clearTooltipText,
  closeOnSelect = true,
  dataTest,
  disabled = false,
  error,
  footer,
  isCompact = false,
  isOpen = false,
  loading,
  multiple = false,
  noResultsText = 'No Results',
  onAfterClose,
  onChange,
  onClear,
  onClose,
  onInput,
  onOpen,
  onScrollEnd,
  options,
  scrollEndThreshold,
  search,
  selectedOption,
  showSearchIcon = true,
  ...rest
}: ComboboxProps) => {
  const comboboxOptions = React.useMemo(
    () =>
      options?.map((option): OptionType => {
        return {
          ...option,
          type: 'list',
        };
      }),
    [options]
  );

  const comboboxRef = React.useRef<HTMLDivElement>(null);
  const searchInputRef = React.useRef<HTMLInputElement>(null);
  const [inputValue, setInputValue] = React.useState('');
  const [currentFocusedOption, setCurrentFocusedOption] = React.useState(
    !Array.isArray(selectedOption) ? selectedOption : undefined
  );
  const [currentSelectedOption, setCurrentSelectedOption] = React.useState(
    !Array.isArray(selectedOption) ? selectedOption : undefined
  );
  const [showMenu, setShowMenu] = React.useState(isOpen);
  const [results, setResults] = React.useState<OptionType[]>(
    comboboxOptions || []
  );

  const keepOpen = !closeOnSelect || multiple;

  // Setting state is asynchronous so these actions
  // absolutely need to happen when the menu is either fully closed or open.
  React.useEffect(() => {
    if (showMenu) {
      // Opening menu actions
      if (searchInputRef && searchInputRef.current) {
        searchInputRef.current.focus({ preventScroll: true });
      }
    }
  }, [options, showMenu]);

  React.useEffect(() => {
    if (comboboxOptions) {
      setResults(comboboxOptions);
    }
  }, [comboboxOptions]);

  const optionText = (option: OptionType | undefined): string | undefined => {
    if (option?.type === 'member') {
      return option?.member;
    }
    return option?.label;
  };

  const isOptionDisabled = (option: OptionType | undefined): boolean => {
    if (option?.type === 'list') {
      return option?.disabled ?? false;
    }
    return false;
  };

  const toggleMenu = (isOpen: boolean, keepFocus = false): void => {
    if (isOpen) {
      // Closing menu actions
      if (searchInputRef && searchInputRef.current) {
        setInputValue('');
      }

      if (keepFocus && comboboxRef && comboboxRef.current) {
        comboboxRef.current.focus();
      }

      if (onClose) {
        onClose();
      }
    } else if (currentFocusedOption !== currentSelectedOption) {
      // This happens when the user doesn't select an option by keyboard.
      setCurrentFocusedOption(currentSelectedOption);
    }

    if (!isOpen && onOpen) {
      onOpen();
    }

    setShowMenu(!isOpen);
  };

  const updateOption = (option: OptionType): void => {
    const hasChanged = option !== currentSelectedOption;

    if (!keepOpen) {
      toggleMenu(true, true);
    }

    if ((!multiple && !hasChanged) || isOptionDisabled(option)) {
      return;
    }

    setCurrentSelectedOption(option);

    if (onChange) {
      onChange(option);
    }
  };

  const hoverOption = (option: OptionType): void => {
    setCurrentFocusedOption(option);
  };

  const focusOption = (direction: FocusDirection = 'first'): void => {
    const options = results.filter((option) => !isOptionDisabled(option));
    if (!options.length) return;

    let currentFocusedIndex = -1;
    if (currentFocusedOption) {
      currentFocusedIndex = options.indexOf(currentFocusedOption);
    } else if (currentSelectedOption) {
      currentFocusedIndex = options.indexOf(currentSelectedOption);
    }

    switch (direction) {
      case 'up':
        setCurrentFocusedOption(
          options[
            currentFocusedIndex > 0
              ? currentFocusedIndex - 1
              : options.length - 1
          ]
        );
        break;
      case 'down':
        setCurrentFocusedOption(
          options[(currentFocusedIndex + 1) % options.length]
        );
        break;
      case 'last':
        setCurrentFocusedOption(options[options.length - 1]);
        break;
      default:
        setCurrentFocusedOption(options[0]);
        break;
    }
  };

  const targetIsClearBtn = (element: HTMLElement): boolean => {
    return !!element.closest('.ids-combobox-input__clear');
  };

  const targetIsActionBtn = (element: HTMLElement): boolean => {
    return !!element.closest('.ids-list-item__text-action');
  };

  const targetIsFooter = (element: HTMLElement): boolean => {
    return !!element.closest('.ids-combobox__footer');
  };

  const resetComboboxFocus = (): void => {
    if (search && searchInputRef && searchInputRef.current) {
      if (document.activeElement !== searchInputRef.current) {
        searchInputRef.current.focus();
      }
    } else if (comboboxRef && comboboxRef.current) {
      if (document.activeElement !== comboboxRef.current) {
        comboboxRef.current.focus();
      }
    }
  };

  const handleOnKeyDown = (
    keyboardEvent: React.KeyboardEvent<HTMLDivElement>
  ): void => {
    const { target } = keyboardEvent;
    switch (keyboardEvent.key) {
      case Keys.Escape:
        if (showMenu) {
          toggleMenu(showMenu);
        }
        break;
      case Keys.Enter:
        if (
          !targetIsClearBtn(target as HTMLElement) &&
          !targetIsActionBtn(target as HTMLElement)
        ) {
          keyboardEvent.preventDefault();
          keyboardEvent.stopPropagation();
          if (currentFocusedOption && showMenu) {
            updateOption(currentFocusedOption);
          }
          if ((!currentFocusedOption && showMenu) || !showMenu) {
            toggleMenu(showMenu, true);
          }
        }
        break;
      case Keys.Space:
        if (!showMenu && !targetIsClearBtn(target as HTMLElement)) {
          keyboardEvent.preventDefault();
          keyboardEvent.stopPropagation();
          toggleMenu(false);
        }
        break;
      case Keys.ArrowUp:
        if (showMenu) {
          keyboardEvent.preventDefault();
          keyboardEvent.stopPropagation();

          resetComboboxFocus();
          focusOption('up');
        }
        break;
      case Keys.ArrowDown:
        if (showMenu) {
          keyboardEvent.preventDefault();
          keyboardEvent.stopPropagation();

          resetComboboxFocus();
          focusOption('down');
        }
        break;
      default:
        break;
    }
  };
  const handleOnClick = (e: React.MouseEvent<HTMLDivElement>): void => {
    const { target } = e;
    if (
      disabled ||
      targetIsClearBtn(target as HTMLDivElement) ||
      targetIsFooter(target as HTMLDivElement)
    ) {
      return;
    }
    toggleMenu(showMenu, true);
  };

  const handleSearch = (searchTerm: string): void => {
    if (onInput) {
      onInput(searchTerm);
    }

    setInputValue(searchTerm);
  };

  const handleClear = (): void => {
    if (showMenu) {
      toggleMenu(showMenu);
    }
    if (multiple && onClear) {
      onClear();
    }
    setCurrentSelectedOption(undefined);
    setCurrentFocusedOption(undefined);
  };

  React.useEffect(() => {
    if (showMenu !== isOpen) {
      toggleMenu(!isOpen);
    }
    // Adding toggleMenu to the dependency array causes unwanted affects
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isOpen]);

  React.useEffect(() => {
    if (currentFocusedOption && results.length) {
      const isFocusedVisible = results.indexOf(currentFocusedOption) >= 0;
      if (!isFocusedVisible) {
        setCurrentFocusedOption(undefined);
      }
    }
  }, [currentFocusedOption, results]);

  const canShowMenu = showMenu && !disabled;

  const comboboxClassname = cx('ids-combobox', className, {
    'ids-combobox--active': canShowMenu,
    'ids-combobox--compact': isCompact,
    'ids-combobox--disabled': disabled,
    'ids-combobox--error': error,
  });

  const comboboxDropdownClassname = cx('ids-combobox__dropdown', {
    [`${className}__dropdown`]: !!className,
    'ids-combobox__dropdown--compact': isCompact,
    'ids-combobox__dropdown--has-footer': !!footer,
  });

  const noResultsOrLoading = loading ? (
    <List
      isCompact
      loading
      multiple={multiple}
      className="ids-combobox__list"
    />
  ) : (
    <div className="ids-combobox__no-results">{noResultsText}</div>
  );

  return (
    <div
      ref={comboboxRef}
      className={comboboxClassname}
      data-test={dataTest}
      onClick={handleOnClick}
      onKeyDown={handleOnKeyDown}
      role="button"
      tabIndex={0}
      {...rest}
    >
      <Dropdown
        key="comboboxDropdown"
        content={
          results && results.length ? (
            <List
              options={results}
              isCompact
              onOptionFocus={hoverOption}
              onOptionChange={updateOption}
              onOptionBlur={() => setCurrentFocusedOption(undefined)}
              selectedOption={multiple ? selectedOption : currentSelectedOption}
              focusedOption={currentFocusedOption}
              multiple={multiple}
              disableTabbing
              className="ids-combobox__list"
            />
          ) : (
            noResultsOrLoading
          )
        }
        footer={footer}
        isScrollable
        isOpen={canShowMenu}
        className={comboboxDropdownClassname}
        isReferenceWidth={!autoWidth}
        onClose={() => {
          toggleMenu(true, true);
        }}
        onAfterClose={onAfterClose}
        onScrollEnd={onScrollEnd}
        scrollEndThreshold={scrollEndThreshold}
      >
        {multiple ? (
          <ComboboxInput
            isOpen={canShowMenu}
            search={search}
            onSearch={handleSearch}
            searchRef={searchInputRef}
            clear={
              selectedOption &&
              Array.isArray(selectedOption) &&
              selectedOption.length
                ? clear
                : undefined
            }
            onClear={handleClear}
            clearTooltipText={clearTooltipText}
            disabled={disabled}
            isPlaceholder
            label={children}
            showSearchIcon={showSearchIcon}
          />
        ) : (
          <ComboboxInput
            isOpen={canShowMenu}
            search={search}
            onSearch={handleSearch}
            searchRef={searchInputRef}
            clear={currentSelectedOption ? clear : undefined}
            onClear={handleClear}
            clearTooltipText={clearTooltipText}
            disabled={disabled}
            icon={currentSelectedOption?.icon}
            isPlaceholder={!currentSelectedOption}
            label={optionText(currentSelectedOption) || children}
            src={currentSelectedOption?.src}
            color={currentSelectedOption?.color}
            showSearchIcon={showSearchIcon}
            value={inputValue}
          />
        )}
      </Dropdown>
    </div>
  );
};

export default Combobox;
