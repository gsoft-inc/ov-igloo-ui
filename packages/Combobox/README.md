# Combobox

A combobox has all the features a select does as well as filtering and clearing results.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/combobox` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/combobox
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/combobox
```

## Usage

Then to use the component in your code just import it!

```jsx
import Combobox from '@igloo-ui/combobox';

const optionList = [
  {
    label: 'Text 1',
    value: '1',
  },
  {
    label: 'Text 2 (disabled)',
    value: '2',
    disabled: true,
  },
  {
    label:
      'Text 3. I will put a lot of text here to see how it behaves. Hopefully it looks good!',
    value: '3',
    color: '#74DCC9',
  },
];

<Combobox options={optionList} search clear clearTooltipText="Remove Mapping">
  Place holder text
</Combobox>;
```
