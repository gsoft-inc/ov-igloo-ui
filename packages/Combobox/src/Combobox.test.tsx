/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import MockDropdown from '@igloo-ui/dropdown/src/__mocks__/Dropdown.mock';
import MockTooltip from '@igloo-ui/tooltip/src/__mocks__/Tooltip.mock';

import Combobox, { ComboboxOption } from './Combobox';

jest.mock('@igloo-ui/dropdown', () => ({
  __esModule: true,
  default: jest.fn(MockDropdown),
}));

jest.mock('@igloo-ui/tooltip', () => ({
  __esModule: true,
  default: jest.fn(MockTooltip),
}));

const setup = (props = {}, options: ComboboxOption[] = []) => {
  options =
    options && options.length > 0
      ? options
      : [
          {
            label: 'Option 1',
            value: '1',
          },
          {
            label: 'Option 2',
            value: '2',
          },
        ];
  return render(
    <Combobox dataTest="combobox1" options={options} {...props}>
      Hello world
    </Combobox>
  );
};

describe('Combobox', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('combobox1');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup({ isOpen: true });
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should render medium size as default', () => {
    setup();
    const combobox = screen.getByTestId('combobox1');

    expect(combobox).not.toHaveClass('ids-combobox--compact');
  });

  test('It should render compact size when set', () => {
    setup({ isCompact: true });
    const combobox = screen.getByTestId('combobox1');

    expect(combobox).toHaveClass('ids-combobox--compact');
  });

  test('It should render in error state when set', () => {
    setup({ error: true });

    const combobox = screen.getByTestId('combobox1');

    expect(combobox).toHaveClass('ids-combobox--error');
  });

  test('It should render in disabled state when set', () => {
    setup({ disabled: true });

    const combobox = screen.getByTestId('combobox1');

    expect(combobox).toHaveClass('ids-combobox--disabled');
  });

  test('It should render in active state when isOpen', () => {
    setup({ isOpen: true });

    const combobox = screen.getByTestId('combobox1');

    expect(combobox).toHaveClass('ids-combobox--active');
  });

  test('It should render a clear button when item is selected and selected item is cleared when clicked', () => {
    setup({ isOpen: true, clear: true });

    const combobox = screen.getByTestId('combobox1');

    const comboboxOptions = combobox.querySelectorAll('.ids-list-item');
    fireEvent.click(comboboxOptions[0]);

    const clearBtn = combobox.querySelector('.ids-combobox-input__clear');
    expect(clearBtn).toBeInTheDocument();

    if (clearBtn) {
      fireEvent.click(clearBtn);
    }
    const comboboxValue = combobox.querySelector('.ids-combobox-input');

    expect(comboboxValue).toHaveClass('ids-combobox-input--placeholder');
  });

  /* Dropdown */

  test('It should display the Dropdown when isOpen is set', () => {
    const { container } = setup({ isOpen: true });

    const dropdown = container.querySelector('.ids-dropdown');

    expect(dropdown).toBeInTheDocument();
  });

  test('It should display the Dropdown when the Combobox is clicked', () => {
    setup();
    const combobox = screen.getByTestId('combobox1');
    fireEvent.click(combobox);

    const dropdown = combobox.querySelector('.ids-dropdown');

    expect(dropdown).toBeInTheDocument();
  });

  test('It should not display the Dropdown when the combobox is clicked if the Combobox is disabled', () => {
    setup({ disabled: true });

    const combobox = screen.getByTestId('combobox1');
    fireEvent.click(combobox);
    const dropdown = combobox.querySelector('.ids-dropdown');

    expect(dropdown).not.toBeInTheDocument();
  });

  /* Options */

  test('It should render all options', () => {
    const options: ComboboxOption[] = [
      { label: '1', value: 1 },
      { label: '2', value: 2 },
      { label: '3', value: 3 },
      { label: '4', value: 4 },
    ];

    setup({ isOpen: true }, options);
    const combobox = screen.getByTestId('combobox1');
    const comboboxOptions = combobox.querySelectorAll('.ids-list-item');

    expect(comboboxOptions.length).toBe(options.length);
  });

  test("It should put the option label in the header after it's clicked", () => {
    const options = [{ label: '1', value: 1 }];

    setup({ isOpen: true }, options);
    const combobox = screen.getByTestId('combobox1');
    const comboboxOptions = combobox.querySelectorAll('.ids-list-item');
    fireEvent.click(comboboxOptions[0]);
    const comboboxValue = combobox.querySelector('.ids-combobox-input');

    expect(comboboxValue).not.toHaveClass('ids-combobox-input--placeholder');
  });

  test("It should not put the option label in the header after it's clicked if the option is disabled", () => {
    const options = [{ label: '1', value: 1, disabled: true }];

    setup({ isOpen: true }, options);
    const combobox = screen.getByTestId('combobox1');
    const comboboxOptions = combobox.querySelectorAll('.ids-list-item');
    fireEvent.click(comboboxOptions[0]);
    const comboboxValue = combobox.querySelector('.ids-combobox-input');

    expect(comboboxValue).toHaveClass('ids-combobox-input--placeholder');
  });

  test('It should render a checkbox if set to multiple', () => {
    setup({ isOpen: true, multiple: true });
    const combobox = screen.getByTestId('combobox1');
    const comboboxCheckbox = combobox.querySelector('.ids-list-item__checkbox');

    expect(comboboxCheckbox).toBeInTheDocument();
  });
});
