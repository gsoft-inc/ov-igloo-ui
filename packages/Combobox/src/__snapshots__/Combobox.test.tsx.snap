// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-combobox ids-combobox--active"
    data-test="combobox1"
    role="button"
    tabindex="0"
  >
    <div
      class="ids-dropdown__ref"
    >
      <div
        class="ids-combobox-input ids-combobox-input--placeholder"
      >
        <div
          class="ids-combobox-input__value"
        >
          <span
            class="ids-combobox-input__value-label"
          >
            Hello world
          </span>
        </div>
        <div
          class="ids-combobox-input__actions"
        >
          <svg
            class="ids-icon ids-combobox-input__chevron"
            fill="none"
            height="16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4.604 10.75c-.891 0-1.337-1.077-.708-1.707l3.397-3.397a1 1 0 0 1 1.414 0l-.497.497.497-.497 3.397 3.397c.63.63.183 1.707-.707 1.707H4.603Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="ids-dropdown ids-combobox__dropdown undefined__dropdown ids-dropdown--bottom-start"
      data-show="true"
      style="position: fixed; top: 0px; left: 0px;"
    >
      <ul
        class="ids-list ids-combobox__list ids-list--compact"
        role="listbox"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="ids-list-item ids-list-item--compact ids-list-item--selectable"
          role="option"
          tabindex="-1"
        >
          <div
            class="ids-list-item__content"
          >
            <span
              class="ids-list-item__text"
            >
              <span
                class="ids-list-item__text-body"
              >
                <span
                  class="ids-list-item__text-content"
                >
                  <span
                    class="ids-list-item__text-label"
                  >
                    Option 1
                  </span>
                </span>
              </span>
            </span>
          </div>
        </li>
        <li
          aria-selected="false"
          class="ids-list-item ids-list-item--compact ids-list-item--selectable"
          role="option"
          tabindex="-1"
        >
          <div
            class="ids-list-item__content"
          >
            <span
              class="ids-list-item__text"
            >
              <span
                class="ids-list-item__text-body"
              >
                <span
                  class="ids-list-item__text-content"
                >
                  <span
                    class="ids-list-item__text-label"
                  >
                    Option 2
                  </span>
                </span>
              </span>
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
